openhtf/util/threads.py

```diff
@@ -92,15 +103,23 @@ def kill(self):
 
   def async_raise(self, exc_type):
     """Raise the exception."""
-    assert self.is_alive(), 'Only running threads have a thread identity'
+    # Should only be called on a started thread so raise otherwise
+    assert self.ident is not None, 'Only started threads have a thread identity'
+
+    # If the thread has died we don't want to raise an exception so log.
+    if not self.is_alive():
+      _LOG.debug('Not raising %s because thread %s (%s) is not alive',
+                 exc_type, self.name, self.ident)
+
     result = ctypes.pythonapi.PyThreadState_SetAsyncExc(
         ctypes.c_long(self.ident), ctypes.py_object(exc_type))
-    if result == 0:
+    if result == 0 and self.is_alive():
       raise ValueError('Thread ID was invalid.', self.ident)
-    elif result != 1:
+    elif result > 1:
       # Something bad happened, call with a NULL exception to undo.
       ctypes.pythonapi.PyThreadState_SetAsyncExc(self.ident, None)
-      raise SystemError('PyThreadState_SetAsyncExc failed.', self.ident)
+      raise RuntimeError('PyThreadState_SetAsyncExc %s failed: %s',
+                         self.ident, result)
```

Comment:
Technically should be something like below, since the command didn't “fail”

```python
'Error: PyThreadState_SetAsyncExc on %s affected %d threads.' % (self.name, result)
```

No one really knows when this case is reached.

As noted above I don't think we should print `self.ident`

Reply:
Used your text, I'm including all the information we have, thread name, exc_type, thread name and result.  These failure conditions are rare so somewhat hard to replicate so we should acquire as much debugging info as possible.